src/magicbox/SuggestionsManager.ts

```diff
@@ -141,73 +143,13 @@ export class SuggestionsManager {
     this.keyboardFocusedElement = null;
   }
 
-  public mergeSuggestions(suggestions: Array<Promise<Suggestion[]> | Suggestion[]>, callback?: (suggestions: Suggestion[]) => void) {
-    let results: Suggestion[] = [];
-    let timeout;
-    let stillNeedToResolve = true;
-    // clean empty / null values in the array of suggestions
-    suggestions = compact(suggestions);
-    const promise = (this.pendingSuggestion = new Promise<Suggestion[]>((resolve, reject) => {
-      // Concat all promises results together in one flat array.
-      // If one promise take too long to resolve, simply skip it
-      each(suggestions, (suggestion: Promise<Suggestion[]>) => {
-        let shouldRejectPart = false;
-        setTimeout(function() {
-          shouldRejectPart = true;
-          stillNeedToResolve = false;
-        }, this.options.timeout);
-        suggestion.then((item: Suggestion[]) => {
-          if (!shouldRejectPart && item) {
-            results = results.concat(item);
-          }
-        });
-      });
-
-      // Resolve the promise when one of those conditions is met first :
-      // - All suggestions resolved
-      // - Timeout is reached before all promises have processed -> resolve with what we have so far
-      // - No suggestions given (length 0 or undefined)
-      const onResolve = () => {
-        if (stillNeedToResolve) {
-          if (timeout) {
-            clearTimeout(timeout);
-          }
-          if (results.length == 0) {
-            resolve([]);
-          } else if (promise == this.pendingSuggestion || this.pendingSuggestion == null) {
-            resolve(results.sort((a, b) => b.index - a.index));
-          } else {
-            reject('new request queued');
-          }
-        }
-        stillNeedToResolve = false;
-      };
-
-      if (suggestions.length == 0) {
-        onResolve();
-      }
-      if (suggestions == undefined) {
-        onResolve();
-      }
-
-      timeout = setTimeout(function() {
-        onResolve();
-      }, this.options.timeout);
-
-      Promise.all(suggestions).then(() => onResolve());
-    }));
-
-    promise
-      .then((suggestions: Suggestion[]) => {
-        if (callback) {
-          callback(suggestions);
-        }
-        this.updateSuggestions(suggestions);
-        return suggestions;
-      })
-      .catch(() => {
-        return null;
-      });
+  public async mergeSuggestions(suggestions: Array<Promise<Suggestion[]> | Suggestion[]>) {
+    const { results, status } = await this.suggestionsProcessor.processQueries(suggestions);
+    if (status === ProcessingStatus.Overriden) {
+      return [];
+    }
+    this.updateSuggestions(results);
```

Comment:
Related to my comment above,
1- What does passing an empty array to this method do?
2- On line 151, we update the suggestions. Can we retrieve these values from the class by exposing a getter rather than returning the results and storing them in the MagicBox class? My feeling right now is this method should be `void`.

Reply:
Could it be easier for others working with the code to expose an `emptySuggestions` or `clearSuggestions` method?

```
this.suggestionsManager.clearSuggestions()
```

When I see `suggestionsManager.mergeSuggestions([])`, my first intuition is that the method will merge what I pass with the rest of the suggestions.

Reply:
Good point, I'll rename the method and add clearSuggestions.

Reply:
UPDATE: this fix was moved to [JSUI-2700](https://coveord.atlassian.net/browse/JSUI-2700).